[TNB-7] Save game stats

Squidfishrl opened the pull request:
## What's here?
- Use Room DB for client side data storage 
- Add a repository class that abstracts away DB specifics
- Save stats after each completed game

## Tickets involved
#7

Squidfishrl committed:
fix package name change discrepencies

Squidfishrl committed:
add repository and room database (with gamestats table)

Squidfishrl committed:
add repository to view model

Squidfishrl committed:
write game stat to db and read most recent game from it

Squidfishrl committed:
follow style guidelines

Squidfishrl committed:
fetch last played game on ViewModel creation

also increase round counter text size in the game screen

Squidfishrl committed:
fix tests

Squidfishrl committed:
add dependency injection

Squidfishrl committed:
comform to linter rules

sivanov-mlb reviewed (commented):
.

Squidfishrl committed:
refactor code

- decouple screen composables from the ViewModel
- remove None enum record for GameModule and Game
- small refactors

Squidfishrl committed:
update tests

Stanislav04 commented on .idea/androidTestResultsUserPreferences.xml:
Not sure if we need this. If it is updated with almost every little interaction with Android Studio, I might check it out of git

Stanislav04 commented on app/src/androidTest/java/org/squidfish/tuplenback/LocalStorageRepositoryTests.kt:
```diff
@@ -0,0 +1,85 @@
+import org.squidfish.tuplenback.games.GameModule
+import org.squidfish.tuplenback.models.GameStatsModel
+
+@RunWith(AndroidJUnit4::class)
```
This is automatically the annotation for tests in the `androidTests` folder.
These tests look simple enough for us to pull on the JVM. Take a look at Robolectric

Stanislav04 commented on app/src/androidTest/java/org/squidfish/tuplenback/LocalStorageRepositoryTests.kt:
```diff
@@ -0,0 +1,85 @@
+@RunWith(AndroidJUnit4::class)
+class LocalStorageRepositoryTests {
+    private lateinit var db: AppDatabase
+    private lateinit var rep: LocalStorageRepository
```
Can we use the full `repository` word?

Stanislav04 commented on app/src/androidTest/java/org/squidfish/tuplenback/LocalStorageRepositoryTests.kt:
```diff
@@ -0,0 +1,85 @@
+
+    @After
+    fun closeDb() {
+        if (::db.isInitialized) {
```
I don't think you need the `::`

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/TupleNBackApplication.kt:
```diff
@@ -4,15 +4,18 @@ import android.app.Application
         startKoin {
             androidContext(this@TupleNBackApplication)
             modules(
+                DatabaseModule().module,
```
Since it's not very likely to have more than one database in this app, maybe it should be `DataSourceModule`? This is just a suggestion

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/BaseGameStats.kt:
```diff
@@ -0,0 +1,21 @@
+import org.squidfish.tuplenback.games.Game
+import org.squidfish.tuplenback.games.GameModule
+
+interface BaseGameStats {
```
Why is this an interface instead of a domain level data class you map to?

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/GameStatsMapper.kt:
```diff
@@ -0,0 +1,29 @@
+import org.squidfish.tuplenback.data.room.GameStatsData
+import org.squidfish.tuplenback.models.GameStatsModel
+
+class GameStatsMapper {
+    companion object {
```
You can make `GameStatsMapper` `object`. Also have you though about implementing mappers as extension functions/properties?

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/GameStatsMapper.kt:
```diff
@@ -0,0 +1,29 @@
+    companion object {
+        fun toData(stats: GameStatsModel): GameStatsData = GameStatsData(
+            gameEndTime = stats.gameEndTime,
+            gameType = stats.gameType ?: throw IllegalArgumentException("Game type is unset"),
+            gameModule = stats.gameModule ?: throw IllegalArgumentException("Game module is unset"),
```
We have to talk about how we handle incorrect states. Do we use `try`/`catch` or a `Result` (though a custom one) class. The problem with `try`/`catch` is that the function signature does not indicate that an error is thrown, therefore you need to jump the source code to see if you need to use a `try`/`catch` block (sure, it can be documented, but still does not provide compile-time safety)

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/LocalStorageRepository.kt:
```diff
@@ -0,0 +1,15 @@
+import org.squidfish.tuplenback.data.room.AppDatabase
+import org.squidfish.tuplenback.models.GameStatsModel
+
+@Single
```
Can we refrain from annotating classes like this? If we decide to swap out Koin, it would have much smaller impact if we need to update only the module files (and the Application's `startKoin` call

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/room/GameStatsDao.kt:
```diff
@@ -0,0 +1,35 @@
+
+@Dao
+interface GameStatsDao {
+    @Query(
+        """
+        SELECT * FROM gamestatsdata
+        WHERE gameType = :gameType
+        ORDER BY gameEndTime DESC
+        LIMIT :amount
+    """,
+    )
```
The `"""` can be pulled to be on the same lines as the braces (I don't think that violates the formatting the linter sets but correct me if I'm wrong)

Squidfishrl replied:
It does, unfortunately. Perhaps we should look into updating the linter rules?

Stanislav04 replied:
Or we can just indent the closing braces as well?

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/di/RepositoryModule.kt:
```diff
@@ -1,6 +1,13 @@
+class RepositoryModule {
+
+    @Single
+    fun createRepository(db: AppDatabase): LocalStorageRepository = LocalStorageRepository(db)
```
You provide the repository manually, so there indeed is no reason to have the class declaration annotated with `@Single`

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameStatsModel.kt:
```diff
@@ -0,0 +1,22 @@
+ */
+data class GameStatsModel(
+    override val gameEndTime: Long = System.currentTimeMillis(),
+    override val gameType: Game? = null,
+    override val gameModule: GameModule? = null,
```
Are you sure you want to allow the existense of `GameStatsModel` with no type/module?

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/LocalStorageRepository.kt:
```diff
@@ -0,0 +1,15 @@
+import org.squidfish.tuplenback.models.GameStatsModel
+
+@Single
+class LocalStorageRepository(val db: AppDatabase) {
```
`private val db`?

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -29,14 +30,24 @@ private const val TAG = "GameViewModel"
+        viewModelScope.launch {
+            val prevGame = repository.getRecent()?.gameType
+
+            if (prevGame != null) {
+                game = prevGame
+            }
```
`prevGame?.let { game = it }` (just a personal preference, I know yours is smart cast)

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -29,14 +30,24 @@ private const val TAG = "GameViewModel"
 
     private val gameEngines = mutableMapOf<GameModule, GameEngine>()
-    var game: Game = Game.None // TODO: load game played on last session
+    var game: Game? = null
```
Should this at least be a `private set`?

Squidfishrl replied:
You're right. It should be fixed now

Stanislav04 replied:
Okay but about the `playerStats` below. It's a publicly accessible `var`

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -66,12 +77,26 @@ class GameViewModel : ViewModel() {
+        viewModelScope.launch {
+            val prevGame = repository.getRecent()?.gameType
+
+            if (prevGame == null) {
+                return@launch
+            }
```
You can just add `?: return@launch` at the end of the `prevGame` assignment

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -102,9 +124,9 @@ class GameViewModel : ViewModel() {
         Log.i(TAG, "Initializing game engines")
 
-        verifyGame()
+        val game = requireNotNull(game) { throw IllegalStateException("Game cannot be null") }
```
Again, we have to set up a broader discussion about errors and exceptions

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -102,9 +124,9 @@ class GameViewModel : ViewModel() {
+        val game = requireNotNull(game) { throw IllegalStateException("Game cannot be null") }
 
-        if (game.modules.isEmpty()) {
+        if (game.modules.isEmpty() == true) {
```
Why?

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -208,16 +237,36 @@ class GameViewModel : ViewModel() {
+
+                        val stats = gameEngine.getStats()
+
+                        gameStatsModel[game]?.let { it1 ->
```
If you are giving the argument a name, can it be more meaningful than `it1`. I guess it's because of shadowing warning so choose one `it` to name properly, it doesn't necessarily have to be the inner one

Stanislav04 replied:
Also notice how even though you perform `== null` check above, it is not smart cast here. Another reason to rethink the `throw`/`null` situation

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -208,16 +237,36 @@ class GameViewModel : ViewModel() {
+            // save stats
+            viewModelScope.launch {
+                for (stats in gameState.value.gameStatsModel) {
+                    (repository as LocalStorageRepository).insert(stats.value)
```
`repository` is of type `LocalStorageRepository`

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/views/GameScreen.kt:
```diff
@@ -17,31 +17,29 @@ import androidx.compose.material3.LinearProgressIndicator
 
 @Composable
 fun GameScreen(
     modifier: Modifier,
```
`modifier` should be the first optional parameter of a composable (no need to change it now, just keep it in mind)

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/views/GameScreen.kt:
```diff
@@ -91,8 +89,8 @@ fun GameScreen(
-                    games = viewModel.game.modules,
-                    onGuess = { viewModel.onEvent(AppEvent.MakeMnemonicRepeatGuess(it)) },
+                    games = game.modules,
+                    onGuess = { onGuess(it) },
```
`onGuess = onGuess`

Squidfishrl replied:
Inexplicably, ```onGuess``` is never called unless I use ```onGuess(it)```

Stanislav04 replied:
That is odd. Will have a look at it at some point

Stanislav04 replied:
If you could add a `TODO` to take a look in the future, thank you

Stanislav04 reviewed (commented):
Left some food for thought and some topics for discussion

Squidfishrl committed:
improve game data transfer

Squidfishrl committed:
major refactors

- Change GameModel. It is now in a one-to-many relation with its DTO
- Return a Result type instead of throwing exceptions
- Update database schema in preparation for game levels
- Views receive leaner dependencies
- Use Robolectric to speedup repository tests
- small fixes and improvements

Stanislav04 commented on app/build.gradle.kts:
```diff
@@ -72,14 +72,22 @@ dependencies {
     androidTestImplementation(libs.androidx.ui.test.junit4)
     debugImplementation(libs.androidx.ui.tooling)
     debugImplementation(libs.androidx.ui.test.manifest)
+    testImplementation(kotlin("test"))
```
Can you move it up with the other `testImplementation`s

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/GameMapper.kt:
```diff
@@ -0,0 +1,71 @@
+import org.squidfish.tuplenback.utils.MissingGameStats
+import org.squidfish.tuplenback.utils.Result
+
+fun List<GameStatsData>.asGameModel(): Result<GameModel, Error> {
```
Since it takes no params, it can be an extension field

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/GameMapper.kt:
```diff
@@ -0,0 +1,71 @@
+        return Result.Error(MissingGameStats)
+    }
+
+    val first = this.first()
```
`val first = firstOrNull() ?: return Result.Error(MissingGameStats)`

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/GameMapper.kt:
```diff
@@ -0,0 +1,71 @@
+        GameModel(
+            gameType = first.gameType,
+            playerStats = playerStats,
+            gameSettings = LevelRepository.getSettings(first.gameType, first.level),
```
I'd rather have some data layer class that holds game type and level and has a mapper for this. Alternatively, inject the repository

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/GameMapper.kt:
No need to explicitly call `this` in the extensions

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/LocalStorageRepository.kt:
```diff
@@ -0,0 +1,46 @@
+                return Result.Success(Unit)
+            }
+            is Result.Error -> {
+                return stats
```
You can omit the `return`s

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/LocalStorageRepository.kt:
```diff
@@ -0,0 +1,46 @@
+                return Result.Success(Unit)
+            }
+            is Result.Error -> {
+                return stats
```
Same as above

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt:
```diff
@@ -26,17 +33,32 @@ private const val TAG = "GameViewModel"
+    init {
+        viewModelScope.launch {
+
+            repository.getRecent().onSuccess {
+                it?.gameType?.let { prevGame -> game = prevGame }
+            }.onError {
+                Log.e(TAG, "Cannot initialize ViewModel. Repository error: $it")
+            }
```
Prefer doing this in the `onStart` of the state flow

Squidfishrl replied:
I'm unsure whether that could work. The flow's onStart method is invoked only after a game has begun. For consistency reasons, it becomes necessary to know the previous game when the "Play recent" button is pressed. If this button is pressed before another game has been started, `onStart` would not be called, and the player would be unable to play, when they, in fact, should.

Stanislav04 replied:
It can stay in the init this time (technically, you're not totally right that it would not be loaded, at least not if everything is structured correctly)

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/PlayerPerformanceStats.kt:
```diff
@@ -0,0 +1,18 @@
+    val missedRecalls: Int = 0,
+    val correctNonRecalls: Int = 0,
+) {
+    fun getCurrentRound(): Int = correctRecalls + incorrectRecalls + missedRecalls + correctNonRecalls
```
Can this be a field `currentRound`?

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/models/Repository.kt:
```diff
@@ -0,0 +1,3 @@
+package org.squidfish.tuplenback.models
+
+interface Repository
```
Still not sure why we need that 😕

Squidfishrl replied:
Because we'd want to decouple the `GameViewModel` from its data package dependencies, which is what `LocalStorageRepository` happens to be. So `LocalStorageRepository` would implement `Repository` and `GameViewModel` would depend on that type, thus removing the data package import. I may have, however, forgotten to do that...

Stanislav04 replied:
The intention is not clear at all because of at least two things:
- The generic name `Repository` which implies any repository would implement it
- It has no requirements making it impossible for anything to depend on it, it becomes just a marker interface

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/utils/Error.kt:
```diff
@@ -0,0 +1,10 @@
+
+object MissingGameStats: Error
+object InconsistentGameStats : Error
+object InconsistentGameData: Error
```
This file should only contain the `Error` interface. Extract these errors in a sealed interface related to the game errors

Stanislav04 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.7-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.14.2-bin.zip
```
I couldn't resolve version 14.2, 11.1 was the latest that worked for me

Stanislav04 reviewed (commented):
Addl' feedback

Squidfishrl committed:
add a repository interface

Includes a WIP level repository and level mapper. Made some small fixes as well

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/LevelMapper.kt:
```diff
@@ -0,0 +1,21 @@
+
+val LevelData.asGameSettings: Result<GameSettings, Error>
+    get() {
+        // val settings = LevelRepository.get(this) - doesn't work because get is a suspend function
```
This is a mapper property, we wouldn't want any data fetching here, fetching is not mapping

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/utils/Error.kt:
```diff
@@ -1,10 +1,3 @@
-object MissingGameStats: Error
-object InconsistentGameStats : Error
-object InconsistentGameData: Error
+sealed interface Error
```
This interface should not be `sealed`

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/utils/DataErrors.kt:
```diff
@@ -0,0 +1,9 @@
+
+object MissingGameStats : ValidationError
+object InconsistentGameStats : ValidationError
+object InconsistentGameData : ValidationError
```
When all the subclasses of a `sealed` interface/class are in the same file with it, prefer placing the subclasses inside the `sealed` parent

Squidfishrl committed:
small fixes

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/GameMapper.kt:
```diff
@@ -0,0 +1,78 @@
+
+        val playerStats = mutableMapOf<GameModule, PlayerPerformanceStats>()
+        forEach { stat ->
+            playerStats[stat.gameModule] = PlayerPerformanceStats(
+                correctRecalls = stat.correctRecalls,
+                incorrectRecalls = stat.incorrectRecalls,
+                missedRecalls = stat.missedRecalls,
+                correctNonRecalls = stat.correctNonRecalls,
+            )
```
This could also be extracted into a mapper field

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/GameMapper.kt:
```diff
@@ -0,0 +1,78 @@
+import org.squidfish.tuplenback.utils.Result
+import org.squidfish.tuplenback.utils.ValidationError
+
+val List<GameStatsData>.asGameModel: Result<GameModel, Error>
```
We have to discuss (not now, but in the future) if we want mapper to return `Result`. It has an advantage, yes, but on the other hand, it makes mapping between model feel like there's more than isomorphism between object, including logic that can fail.

Update: Actually, there should be an extension to our `Result` type that would return the data if present and `null` otherwise, so it's basically like a normal mapper (that has logic that can fail). In other words, I'm fine with this (I don't know how much we would use the information from the error case, but it's a plus to have it now instead of needing code changes later)

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/data/game/GameRepository.kt:
```diff
@@ -0,0 +1,42 @@
+        return recent.asGameModel
+    }
+
+    override suspend fun insert(gameModel: GameModel): Result<Unit, Error> =
+        when (val stats = gameModel.asGameStatsData) {
```
My very subjective opinion - I hate `=` followed by a new line 😄

Stanislav04 commented on app/src/main/java/org/squidfish/tuplenback/utils/GameErrors.kt:
I don't think this is supposed to be in the `utils` package
